File: specs/ts/triangle-wave.ts
import { Spec } from '@uwdata/mosaic-spec';

export const spec : Spec = {
  "meta": {
    "title": "Triangle Wave",
    "description": "A test specification to compare M4 optimized and unoptimized line charts.\n"
  },
  "data": {
    "wave": {
      "file": "data/triangle-wave-day.csv"
    }
  },
  "vconcat": [
    {
      "plot": [
        {
          "mark": "lineY",
          "data": {
            "from": "wave"
          },
          "x": "time_stamp",
          "y": "power",
          "z": null,
          "stroke": "time_stamp"
        },
        {
          "select": "intervalX",
          "as": "$brush"
        }
      ],
      "xLabel": null,
      "width": 680,
      "height": 150
    },
    {
      "vspace": 5
    },
    {
      "plot": [
        {
          "mark": "lineY",
          "data": {
            "from": "wave",
            "filterBy": "$brush"
          },
          "x": "time_stamp",
          "y": "power",
          "z": null,
          "stroke": "time_stamp"
        }
      ],
      "yDomain": "Fixed",
      "colorDomain": "Fixed",
      "xLabel": null,
      "width": 680,
      "height": 150
    },
    {
      "vspace": 5
    },
    {
      "plot": [
        {
          "mark": "lineY",
          "data": {
            "from": "wave",
            "filterBy": "$brush",
            "optimize": false
          },
          "x": "time_stamp",
          "y": "power",
          "z": null,
          "stroke": "time_stamp"
        }
      ],
      "yDomain": "Fixed",
      "colorDomain": "Fixed",
      "xLabel": null,
      "width": 680,
      "height": 150
    }
  ]
};

File: specs/python/triangle-wave.py
import vgplot as vg

meta = vg.meta(
    title="Triangle Wave",
    description="A test specification to compare M4 optimized and unoptimized line charts.\n",
)
wave = vg.csv("data/triangle-wave-day.csv")

brush = vg.selection.intersect()

view = vg.vconcat(
    vg.plot(
        vg.line_y(wave, x="time_stamp", y="power", z=None, stroke="time_stamp"),
        vg.interval_x(bind=brush),
        vg.x_label(None),
        vg.width(680),
        vg.height(150),
    ),
    vg.vspace(5),
    vg.plot(
        vg.line_y(
            data="wave",
            filter_by=brush,
            x="time_stamp",
            y="power",
            z=None,
            stroke="time_stamp",
        ),
        vg.y_domain("Fixed"),
        vg.color_domain("Fixed"),
        vg.x_label(None),
        vg.width(680),
        vg.height(150),
    ),
    vg.vspace(5),
    vg.plot(
        vg.line_y(
            data="wave",
            filter_by=brush,
            optimize=False,
            x="time_stamp",
            y="power",
            z=None,
            stroke="time_stamp",
        ),
        vg.y_domain("Fixed"),
        vg.color_domain("Fixed"),
        vg.x_label(None),
        vg.width(680),
        vg.height(150),
    ),
)

spec = vg.spec()
